### PR TITLE
Update dex contracts

### PR DESCRIPTION
**Pull request by anxolin**
Upgrades dex-contract to their last version.

For now, they don't publish their typescript definitions, so we still have our own.
Also, typechain still doesn't generate the types right, it does a better job though. Now at least it uses string instead og BN.

This PR:
* [X] Updates the contracts to their last version
* [X] Adds a script to generate the typescript definition automatically from the abis

The new structure for the `src/contracts` is:
* `abi`: All the Abis. for now only BatchExchange and ERC20
*  `gen`: autogenerated types
* `BatchExchangeContract.d.ts`: Definition for the BatchExchange contract. Extends the autogenerated types, and redefines the events of it (the part that still not work right in typeschain types)
* `Erc20Contract.d.ts`: Same as above, but for Erc20 contract
* `types.d.ts` aditional types definitions. Used for getting right the events.
* `index.ts`: convenient factory for constructs one instance of the contracts

**test**: 
Checkout the sandboxes. They have instructions on how to run them

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,7 +19,8 @@
     "lint": "eslint src --ext .ts,.js",
     "lint:fix": "eslint --fix . --ext .ts,.js",
     "verify": "npm run lint && npm run test",
-    "sandbox": "DEBUG=ERROR-*,WARN-*,INFO-*,DEBUG-*,-DEBUG-helper*,-DEBUG-util* nodemon -r dotenv/config -r tsconfig-paths/register"
+    "sandbox": "DEBUG=ERROR-*,WARN-*,INFO-*,DEBUG-*,-DEBUG-helper*,-DEBUG-util* nodemon -r dotenv/config -r tsconfig-paths/register",
+    "contracts": "rimraf \"src/contracts/gen\" && npx typechain --target=web3-v1 --outDir src/contracts/gen \"src/contracts/abi/*.json\" && prettier --write src/contracts/gen/*.ts"
```
Comment (Velenir) [inline package.json:23]: `\"src/contracts/abi/*.json\"` Was it not picking up without quotes?

File: src/contracts/Erc20Contract.ts [hunk 2]
```diff
@@ -1,29 +1,29 @@
...
+  value: string
+}
 
-    transferFrom(
-      from: string,
-      to: string,
-      value: number | string | BN,
-    ): TransactionObject<boolean, [string, string, string | number | BN]>
+export interface Erc20Contract extends Omit<Erc20, 'events'> {
+  // Redefine "events" because it's not correctly generated using typechain
```
Comment (Velenir) [inline src/contracts/Erc20Contract.ts:25]: I don't immediately see what is wrong with typechain events. Is it only `allEvents`?

**Review by Velenir — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (anxolin) [reply package.json:23]: for some reason, it was only getting 1 ABI without that, I don't know why...

File: src/contracts/Erc20Contract.ts (hunk 2 above)
Comment (anxolin) [reply src/contracts/Erc20Contract.ts:25]: Notice how I import `ContractEvent` from `types` and not `'./gen/types'`.

They define `EventOptions` and I get it from `web3-eth-contract`

But I might have something else wrong, cause I'm getting an error when I try to use this package, and it's for the events.

Comment (Velenir) [reply src/contracts/Erc20Contract.ts:25]: That's unfortunate. Can you create an issue in their repo

**Review by alfetopito — APPROVED** [review]

File: src/contracts/Erc20Contract.ts (hunk 2 above)
Comment (anxolin) [reply src/contracts/Erc20Contract.ts:25]: I don't have capacity right now, I've created this to create them a PR or to open an issue https://github.com/gnosis/dex-js/issues/31

**Pull request merged**